## Part 1: Explore the Data

Import the data and use Pandas to learn more about the dataset.

In [3]:
import pandas as pd

df = pd.read_csv('Resources/client_dataset.csv')
df.head()

first      last                    job               phone  \
0     Donald   Harding           Immunologist  793-904-7725x39308   
1    Tiffany     Myers        Music therapist    201.442.4543x942   
2    Shannon    Watson           Immunologist   687.737.9424x8503   
3     Nathan     Baker  Accounting technician    827-788-8123x012   
4  Christina  Schwartz           Chiropractor        265-829-3643   

                                 email  client_id  order_id  order_date  \
0     harding.donald.7185@sullivan.com      58515   8953482  2023-04-28   
1    myers.t.6537@ferguson-johnson.net      37609   8069089  2023-05-19   
2                swatson8146@payne.net      57113   1902144  2023-01-29   
3               bakernathan@benson.com      46554   9031802  2023-04-25   
4  christinaschwartz9252@mcconnell.com      92089   1322274  2023-05-28   

   order_week  order_year         item_id     category         subcategory  \
0          17        2023  EUD29711-63-6U        decor            wall art   
1          20        2023  XDA18116-89-4A  consumables                pens   
2           4        2023  ABE59463-05-7E     software  project management   
3          17        2023  ZMM00836-65-0C  consumables                pens   
4          21        2023  BZX55559-12-3X  consumables                misc   

   unit_price  unit_cost  unit_weight  qty  line_number  
0     1096.80     762.71         7.50  105            1  
1       24.95      15.09         1.49   21            0  
2       13.52       7.86         1.68   39            6  
3       36.42      24.85         1.23   29            3  
4      195.10     108.17        46.43   20            1

In [4]:
# View the column names in the data
df.columns

Index(['first', 'last', 'job', 'phone', 'email', 'client_id', 'order_id',
       'order_date', 'order_week', 'order_year', 'item_id', 'category',
       'subcategory', 'unit_price', 'unit_cost', 'unit_weight', 'qty',
       'line_number'],
      dtype='object')

In [5]:
# Use the describe function to gather some basic statistics
df.describe()

client_id      order_id    order_week    order_year    unit_price  \
count  54639.000000  5.463900e+04  54639.000000  54639.000000  54639.000000   
mean   54837.869416  5.470190e+06     11.359139   2022.993064    136.267207   
std    25487.438231  2.599807e+06      7.023499      0.082997    183.873135   
min    10033.000000  1.000886e+06      1.000000   2022.000000      0.010000   
25%    33593.000000  3.196372e+06      6.000000   2023.000000     20.800000   
50%    53305.000000  5.496966e+06     11.000000   2023.000000     68.310000   
75%    78498.000000  7.733869e+06     17.000000   2023.000000    173.160000   
max    99984.000000  9.998480e+06     52.000000   2023.000000   1396.230000   

          unit_cost   unit_weight           qty   line_number  
count  54639.000000  54639.000000  5.463900e+04  54639.000000  
mean      99.446073      5.004116  5.702646e+02      2.979667  
std      133.164267      5.326599  1.879552e+04      2.436320  
min        0.010000      0.000000  0.000000e+00      0.000000  
25%       14.840000      1.450000  3.200000e+01      1.000000  
50%       49.890000      3.240000  6.800000e+01      3.000000  
75%      125.570000      6.890000  1.700000e+02      5.000000  
max      846.270000     46.430000  3.958244e+06      9.000000

In [6]:
# Use this space to do any additional research
# and familiarize yourself with the data.
df.info()


<class 'pandas.core.frame.DataFrame'>
RangeIndex: 54639 entries, 0 to 54638
Data columns (total 18 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   first        54639 non-null  object 
 1   last         54639 non-null  object 
 2   job          54639 non-null  object 
 3   phone        54639 non-null  object 
 4   email        54639 non-null  object 
 5   client_id    54639 non-null  int64  
 6   order_id     54639 non-null  int64  
 7   order_date   54639 non-null  object 
 8   order_week   54639 non-null  int64  
 9   order_year   54639 non-null  int64  
 10  item_id      54639 non-null  object 
 11  category     54639 non-null  object 
 12  subcategory  54639 non-null  object 
 13  unit_price   54639 non-null  float64
 14  unit_cost    54639 non-null  float64
 15  unit_weight  54639 non-null  float64
 16  qty          54639 non-null  int64  
 17  line_number  54639 non-null  int64  
dtypes: float64(3), int64(6), object(9)
memory usag

In [7]:
# What three item categories had the most entries?
df["category"].value_counts().head(3)

category
consumables    23538
furniture      11915
software        8400
Name: count, dtype: int64

In [8]:
# For the category with the most entries, which subcategory had the most entries?
df.loc[df["category"]== "consumables","subcategory"].value_counts().head(1)

subcategory
bathroom supplies    6424
Name: count, dtype: int64

In [9]:
# Which five clients had the most entries in the data?
df["client_id"].value_counts().head(5)

client_id
33615    220
66037    211
46820    209
38378    207
24741    207
Name: count, dtype: int64

In [10]:
# Store the client ids of those top 5 clients in a list.
top_dogs = df["client_id"].value_counts().head(5)
print (top_dogs)

client_id
33615    220
66037    211
46820    209
38378    207
24741    207
Name: count, dtype: int64


In [11]:
# How many total units (the qty column) did the client with the most entries order order?
df.loc[df["client_id"]== 33615,"qty"].sum()

64313

## Part 2: Transform the Data
Do we know that this client spent the more money than client 66037? If not, how would we find out? Transform the data using the steps below to prepare it for analysis.

In [12]:
# Create a column that calculates the subtotal for each line using the unit_price and the qty
df["line_subtotal"]= df["unit_price"] * df["qty"]
df[["unit_price","qty","line_subtotal"]].head(2)

unit_price  qty  line_subtotal
0     1096.80  105      115164.00
1       24.95   21         523.95

In [13]:
# Create a column for shipping price.
# Assume a shipping price of $7 per pound for orders over 50 pounds and $10 per pound for items 50 pounds or under.
# quantity * weight * shipping price 
# if weight <= 50 then shipping price = 10 else spipping price = 7
df["shipping_price"] = df.apply(
    lambda row: row["qty"] * row["unit_weight"] * (7 if row["qty"] * row["unit_weight"] > 50 else 10),
    axis=1
)

# Display the first few rows to verify
df[["qty", "unit_weight", "shipping_price"]].head()


qty  unit_weight  shipping_price
0  105         7.50         5512.50
1   21         1.49          312.90
2   39         1.68          458.64
3   29         1.23          356.70
4   20        46.43         6500.20

In [14]:
# Create a column for the total price using the subtotal and the shipping price along with a sales tax of 9.25%
#logic: total = subtotal *1.095 + shipping 
df["total_price"] = df["line_subtotal"] * 1.0925 + df["shipping_price"]

# Display the first few rows of the new column to verify
df[["line_subtotal", "shipping_price", "total_price"]].head()


line_subtotal  shipping_price    total_price
0      115164.00         5512.50  131329.170000
1         523.95          312.90     885.315375
2         527.28          458.64    1034.693400
3        1056.18          356.70    1510.576650
4        3902.00         6500.20   10763.135000

In [15]:
# Create a column for the cost of each line using unit cost, qty, and
# shipping price (assume the shipping cost is exactly what is charged to the client).
#logic: net_cost = cost*quantity + shipping price  

# Calculate the line cost using unit cost, quantity, and shipping price
df["line_cost"] = (df["unit_cost"] * df["qty"]) + df["shipping_price"]

# Display the first few rows of the new column to verify
df[["unit_cost", "qty", "shipping_price", "line_cost"]].head()



unit_cost  qty  shipping_price  line_cost
0     762.71  105         5512.50   85597.05
1      15.09   21          312.90     629.79
2       7.86   39          458.64     765.18
3      24.85   29          356.70    1077.35
4     108.17   20         6500.20    8663.60

In [16]:
# Create a column for the profit of each line using line cost and line price

# profit for each line 


df["line_profit"] = df["total_price"] - df["line_cost"]

# Display 
df[["total_price", "line_cost", "line_profit"]].head()




total_price  line_cost   line_profit
0  131329.170000   85597.05  45732.120000
1     885.315375     629.79    255.525375
2    1034.693400     765.18    269.513400
3    1510.576650    1077.35    433.226650
4   10763.135000    8663.60   2099.535000

## Part 3: Confirm your work
You have email receipts showing that the total prices for 3 orders. Confirm that your calculations match the receipts. Remember, each order has multiple lines.

Order ID 2742071 had a total price of \$152,811.89

Order ID 2173913 had a total price of \$162,388.71

Order ID 6128929 had a total price of \$923,441.25


In [18]:
# Check your work using the totals above

order_totals = df.groupby('order_id')['total_price'].sum()


print("Calculated total prices for each order:")
print(order_totals)

receipts = {
    2742071: 152811.89,
    2173913: 162388.71,
    6128929: 923441.25
}

print("\nComparison with receipt totals:")
for order_id, expected_total in receipts.items():
    calculated_total = order_totals.get(order_id, None)
    if calculated_total is not None:
        if abs(calculated_total - expected_total) < 0.1:  # Allow small floating-point differences
            print(f"Order {order_id} matches the receipt total: {calculated_total:.2f}")
        else:
            print(f"Order {order_id} does NOT match the receipt total. Calculated: {calculated_total:.2f}, Expected: {expected_total}")
    else:
        print(f"Order {order_id} not found in the data.")
# i think there is a floaring/rounding error in  my code somewhere the falues arent that off

Calculated total prices for each order:
order_id
1000886    1.063291e+05
1001431    4.049920e+05
1003003    1.239231e+04
1003666    7.609830e+05
1004537    1.235545e+05
               ...     
9995243    2.052406e+06
9995530    8.448361e+03
9997244    1.481729e+05
9997426    1.058371e+03
9998480    4.940973e+04
Name: total_price, Length: 10000, dtype: float64

Comparison with receipt totals:
Order 2742071 does NOT match the receipt total. Calculated: 151695.03, Expected: 152811.89
Order 2173913 does NOT match the receipt total. Calculated: 158762.83, Expected: 162388.71
Order 6128929 does NOT match the receipt total. Calculated: 908947.33, Expected: 923441.25


## Part 4: Summarize and Analyze
Use the new columns with confirmed values to find the following information.

In [19]:
# How much did each of the top 5 clients by quantity spend? Check your work from Part 1 for client ids.


top_dogs = df["client_id"].value_counts().head(5).index.tolist()

top_clients_df = df[df["client_id"].isin(top_dogs)]

top_clients_spending = top_clients_df.groupby('client_id')['total_price'].sum()

print(top_clients_spending)



client_id
24741    8.179470e+07
33615    8.208127e+06
38378    1.258933e+07
46820    9.595660e+06
66037    1.013046e+07
Name: total_price, dtype: float64


In [24]:
# Create a summary DataFrame showing the totals for the for the top 5 clients with the following information:
# total units purchased, total shipping price, total revenue, and total profit. 
# Group by 'client_id' and sum the 'qty' to find the top 5 clients by total quantity
client_totals = df.groupby('client_id')['qty'].sum()

#  Sort the clients by total quantity in descending order and select the top 5 clients
top_5_clients = client_totals.sort_values(ascending=False).head(5).index.tolist()

#  Filter the DataFrame to include only the top 5 clients
top_clients_df = df[df['client_id'].isin(top_5_clients)]

# Group by 'client_id' and calculate the totals
summary_df = top_clients_df.groupby('client_id').agg(
    total_units_purchased=('qty', 'sum'),
    total_shipping_price=('shipping_price', 'sum'),
    total_revenue=('total_price', 'sum'),
    total_profit=('line_profit', 'sum')
).reset_index()

#  Display the summary DataFrame
print(summary_df)





   client_id  total_units_purchased  total_shipping_price  total_revenue  \
0      30143                 846069            8981905.24   3.249802e+07   
1      77354                 506382           23814685.86   6.973979e+07   
2      78965                4033302            5340756.88   6.554297e+08   
3      85931                 644423           10122549.19   2.364807e+08   
4      86721                1010426           64335727.72   1.753152e+08   

   total_profit  
0  1.041956e+07  
1  1.022940e+07  
2  2.468588e+08  
3  1.083578e+08  
4  3.702364e+07  


In [29]:
# Format the data and rename the columns to names suitable for presentation.

# Define the money columns. 


# Define a function that converts a dollar amount to millions.


# Apply the currency_format_millions function to only the money columns. 


# Rename the columns to reflect the change in the money format. 

# Define the money columns
money_columns = ['total_shipping_price', 'total_revenue', 'total_profit']

# Define a function that converts a dollar amount to millions
def currency_format_millions(x):
    return x / 1_000_000

# Apply the function to the money columns
if all(col in summary_df.columns for col in money_columns):
    summary_df[money_columns] = summary_df[money_columns].applymap(currency_format_millions)
else:
    print("Some columns are missing. Check column names.")

# Rename the columns to reflect the change in money format
summary_df = summary_df.rename(columns={
    'total_shipping_price': 'Total Shipping Price (in millions)',
    'total_revenue': 'Total Revenue (in millions)',
    'total_profit': 'Total Profit (in millions)',
    'total_units_purchased': 'Total Units Purchased'
})

# Display the formatted DataFrame
print(summary_df)


Some columns are missing. Check column names.
   client_id  Total Units Purchased  Total Shipping Price (in millions)  \
0      30143                 846069                            8.981905   
1      77354                 506382                           23.814686   
2      78965                4033302                            5.340757   
3      85931                 644423                           10.122549   
4      86721                1010426                           64.335728   

   Total Revenue (in millions)  Total Profit (in millions)  
0                    32.498024                   10.419560  
1                    69.739793                   10.229402  
2                   655.429650                  246.858822  
3                   236.480666                  108.357793  
4                   175.315242                   37.023641  


In [35]:
## Sort by 'Total Profit (in millions)' from highest to lowest
sorted_summary_df = summary_df.sort_values('Total Profit (in millions)', ascending=False)

# Check the sorted DataFrame
print(sorted_summary_df)



   client_id  Total Units Purchased  Total Shipping Price (in millions)  \
2      78965                4033302                            5.340757   
3      85931                 644423                           10.122549   
4      86721                1010426                           64.335728   
0      30143                 846069                            8.981905   
1      77354                 506382                           23.814686   

   Total Revenue (in millions)  Total Profit (in millions)  
2                   655.429650                  246.858822  
3                   236.480666                  108.357793  
4                   175.315242                   37.023641  
0                    32.498024                   10.419560  
1                    69.739793                   10.229402  
